brian2/codegen/generators/cpp_generator.py

```diff
@@ -107,7 +107,18 @@ def c_data_type(dtype):
         }}
         '''.format(hightype=hightype, xtype=xtype, ytype=ytype, expr=expr)
 
-_universal_support_code = deindent(mod_support_code)+'''
+floordiv_support_code = ''
+for ix, xtype in enumerate(typestrs):
+    for iy, ytype in enumerate(typestrs):
+        hightype = typestrs[max(ix, iy)]
+        floordiv_support_code += '''
+        {hightype} _brian_floordiv({xtype} x, {ytype} y)
+        {{
+            return ({hightype})floor(1.0*x/y);
```

Comment:
Can we not do better than this by taking into account whether or not x and y are integer types (similar to how we do it for mod just above)? Also, should be inlined I think.

Reply:
Unfortunately at this point we can't, we don't have this information here. We only have it when we do the BAST parsing for optimization... With a refactoring, this would be possible, but I'm not sure we should get into this right now. Another possibility: forget about the negative floor division for now and add it as a known issue -- I think this will be a rather uncommon feature to use.

> (similar to how we do it for mod just above)

Not sure what you mean by that -- mod also introduces an extra function `_brian_mod`.

> Also, should be inlined I think.

As a lesson from that most annoying debugging session of my life, I'm now very hesitant to add `inline` for hypothetical performance benefits :)

Reply:
> The C++ compiler will though. Look at the definition of mod above: it creates a different expression depending if the dtypes are integer or float types.

Ok, I now get what you mean. But what could we improve here -- get rid of the 1.0 multiplication if one of the values is already a floating point variable? I think the compiler would easily optimize this away.

> And for inline, we have it on for mod and I don't think it's causing problems? I would guess that for very simple functions like this it should be OK to inline?

There is something odd about the linking and inline functions (remember my horror story from #928), even though I never quite understood why it would be an issue with C++. So I'd say rather `static inline` or no inline at all (same for `_brian_mod`, just to be sure) -- I have the strong feeling that the compiler will make a smarter decision than we do anyway.

Reply:
> I think we could get rid of the conversion to float and call to floor if both arguments are integer types which would be a big saving. Or am I missing something?

This would give the "truncate to 0" instead of "floor" semantics for negative values, unfortunately. We could add an if statement -- not sure whether this is faster, though?

Happy to use `static inline`, of course.

Reply:
It would, but we could use ``a//b==(a-_brian_mod(a, b))/b``, and ``_brian_mod`` doesn't need to convert to float and then back to int (the latter is apparently particularly slow but I didn't check).

Reply:
Hmm, we could actually use this expression for floating point values as well, no? Maybe avoiding the `floor` function call makes sense there too? Only one way to find out: benchmarking :)

Reply:
In principle yes, but I assume that ``floor`` is less expensive than ``fmod`` and ``_brian_mod`` on floats calls ``fmod`` twice.

Reply:
The verdict is pretty clear...
Here's my test code:
```Python
set_device('cpp_standalone', clean=True)
seed(4321)
G = NeuronGroup(10000000, '''x : 1
                             y : 1
                             a : integer
                             b : integer
                             fvalue : 1
                             ivalue : 1''',
                dtype={'x': np.float, 'y': np.double,
                       'a': np.int32, 'b': np.int64})
G.fvalue = 'rand()*100000'
G.ivalue = 'int(rand()*100000)'
G.run_regularly('b = ivalue//10', name='b')
G.run_regularly('a = ivalue//10', name='a')
G.run_regularly('y = fvalue//10', name='y')
G.run_regularly('x = fvalue//10', name='x')
run(defaultclock.dt, profile=True)
print(profiling_summary())
```
If I run this with `floor(1.0*x/y)` as implementation for `_brian_floordiv`, the codeobjects each take about 10ms (a bit less for the 32bit integer, a bit more for the others). If I run it with `(x-_brian_mod(x, y))/y`, it takes about... 550ms! It seems the `floor` function is quite efficient after all.
As a side note, if we required C++11 (as Brian2GeNN already does), we wouldn't have to loop over the type combinations but could simply write:
```C++
template <typename T, typename U>
static inline typename std::common_type<T, U>::type _brian_floordiv(T x, U y)
{
    return floor(1.0*x/y);
}
```

Reply:
Oh, and removing the 1.0 multiplication for floating point values and just using `x/y` for integer values does not change the runtime in any measurable way for me -- maybe a preference is not needed after all? But there is quite a bit of fluctuation, there might be a small effect.

Reply:
My mind is blown!

The test above is not quite right because ``ivalue`` should be an integer, but the conclusion is the same. With ``ivalue`` an integer the ``a`` and ``b`` codeobjects are now taking twice as long instead of 10x as long.

And now that we know how fast ``floor`` is on floats, I revisited ``_brian_mod`` and it turns out replacing with the single function ``({hightype})(x-y*floor(1.0*x/y))`` is faster in all cases (including the integer case). So I propose we check that it gives the same values and replace.

Reply:
Oh, sorry about `ivalue`. I didn't even think about replacing `_brian_mod` as well but that makes of course perfect sense! What do you think about replacing both functions with a single template function and requiring C++11? If I am not mistaken, then the feature we need is "type traits" (for `std::common_type`, and this has been supported by all compilers for a long time (e.g. for gcc since 4.3 and for MSVC since 14.0, i.e. Visual Studio 2005).

Reply:
I think the only reason we didn't require C++11 already is that Python 2 on Windows is built with MSVC 2008. I have to say I assumed that since C++11 has 11 in the name, it wasn't supported in MSVC 2008 but if that's not the case I would have no objection to using that instead! Would generate much nicer code than the horrible situation we have at the moment.